crates/prql-compiler/src/semantic/resolver/mod.rs

```diff
@@ -982,16 +958,89 @@ impl Resolver {
     }
 
     fn fold_ty_or_expr(&mut self, ty_or_expr: Option<TyOrExpr>) -> Result<Option<TyOrExpr>> {
-        Ok(match ty_or_expr {
+        self.context.root_mod.shadow(NS_THIS);
+        self.context.root_mod.shadow(NS_THAT);
+
+        let res = match ty_or_expr {
             Some(TyOrExpr::Expr(ty_expr)) => {
                 Some(TyOrExpr::Ty(self.fold_type_expr(Some(ty_expr))?.unwrap()))
             }
             _ => ty_or_expr,
-        })
+        };
+
+        self.context.root_mod.unshadow(NS_THIS);
+        self.context.root_mod.unshadow(NS_THAT);
+        Ok(res)
+    }
+}
+
+fn extract_partial_application(mut func: Func, position: usize) -> Func {
+    // Input:
+    // Func {
+    //     params: [x, y, z],
+    //     args: [
+    //         x,
+    //         Func {
+    //             params: [a, b],
+    //             args: [a],
+    //             body: arg_body
+    //         },
+    //         z
+    //     ],
+    //     body: parent_body
+    // }
+
+    // Output:
+    // Func {
+    //     params: [b],
+    //     args: [],
+    //     body: Func {
+    //         params: [x, y, z],
+    //         args: [
+    //             x,
+    //             Func {
+    //                 params: [a, b],
+    //                 args: [a, b],
+    //                 body: arg_body
+    //             },
+    //             z
+    //         ],
+    //         body: parent_body
+    //     }
+    // }
```

Comment:
This should be a snapshot test rather than a comment.

Reply:
This is not something that compiles, it a note about what's happening. Quite verbose one.

Reply:
(this was part of the previous commit, that this one was based on)

Reply:
Huh I'm slightly confused why this showed up in the diff ... oh yeah this should probably say "Input (simplified)" and "Output (simplified)" ...  but yeah that's off-topic for this PR.